guest/x86/bzimage_stub.S

```diff
@@ -0,0 +1,246 @@
+#define MEM_1M             (1024 * MEM_1K)
+#define SECTOR_SIZE        (MEM_1K / 2)
+
+#define ZEROPAGE_MAGIC_SIG    0x53726448    /* "HdrS" in uint32_t */
+#define ZEROPAGE_SETUP_SECTS  0x1f1
+#define ZEROPAGE_HEADER       0x202
+#define ZEROPAGE_CMDLINE_PTR  0x228
+#define ZEROPAGE_IMAGE_SIZE   0x3fc
+#define ZEROPAGE_BINARY_BASE  (4 * SECTOR_SIZE)
+
+#define MBI_FLAGS             0x0
+#define MBI_MEM_LOWER         0x4
+#define MBI_MEM_UPPER         0x8
+#define MBI_CMDLINE           0x10
+#define MBI_MODS_COUNT        0x14
+
+#define MULTIBOOT_HEADER_MAGIC          0x1badb002
+#define MULTIBOOT_INFO_MAGIC		0x2badb002
+#define MULTIBOOT_INFO_HAS_MEMORY	0x00000001
+#define MULTIBOOT_INFO_HAS_BOOT_DEVICE	0x00000002
+#define MULTIBOOT_INFO_HAS_CMDLINE	0x00000004
+#define MULTIBOOT_INFO_HAS_MODS		0x00000008
+#define MULTIBOOT_INFO_HAS_AOUT_SYMS	0x00000010
+#define MULTIBOOT_INFO_HAS_ELF_SYMS	0x00000020
+#define MULTIBOOT_INFO_HAS_MMAP		0x00000040
+#define MULTIBOOT_INFO_HAS_DRIVES	0x00000080
+#define MULTIBOOT_INFO_HAS_CONFIG_TABLE	0x00000100
+#define MULTIBOOT_INFO_HAS_LOADER_NAME	0x00000200
+#define MULTIBOOT_INFO_HAS_APM_TABLE	0x00000400
+#define MULTIBOOT_INFO_HAS_VBE		0x00000800
+
+#define UNIT_TEST_LOAD_ADDRESS     (4 * MEM_1M)
+#define MBI_ADDRESS                (UNIT_TEST_LOAD_ADDRESS - MEM_4K)
+#define CMDLINE_ADDRESS            (UNIT_TEST_LOAD_ADDRESS - MEM_2K)
+
+#define UART_BASE          0x3f8
+#define UART_RBR           0x00
+#define UART_THR           0x00
+#define UART_DLL           0x00
+#define UART_IER           0x01
+#define UART_DLM           0x01
+#define UART_IIR           0x02
+#define UART_FCR           0x02
+#define UART_LCR           0x03
+#define UART_MCR           0x04
+#define UART_LSR           0x05
+#define UART_MSR           0x06
+#define UART_SCR           0x07
+
+	/* Register allocation
+	 *
+	 *     EAX: General, used for IN/OUT and as temporary variables
+	 *     EBX: General, used as a temporary variable
+	 *     ECX: General, used as iterators
+	 *     EDX: Used for IN/OUT when initializing UART and as the zero page pointer afterwards
+	 *     EBP: Used as the base of this stub before jumping to the trampoline, and bzimage pointer afterwards
+	 *     ESI: Used as the source iterator during copying
+	 *     EDI: Used as the destination iterator during copying
+	 */
+
+	.code32
+
+	/* Entry of the stub. Shall be the third sector in the bzimage. */
+start:
+	mov	$0xa0000, %esp
+
+	/* Init UART FIFOs */
+	mov	$(UART_BASE + UART_FCR), %dx
+	mov	$0x7, %al
+	outb	%al, %dx
+
+	/* Init UART data bits / parity bits / stop bits */
+	mov	$(UART_BASE + UART_LCR), %dx
+	mov	$0x3, %al
+	outb	%al, %dx
+
+	/* Disable UART interrupts */
+	mov	$(UART_BASE + UART_IER), %dx
+	mov	$0, %al
+	outb	%al, %dx
+
+	/* Set UART terminal ready and request to send */
+	mov	$(UART_BASE + UART_MCR), %dx
+	mov	$3, %al
+	outb	%al, %dx
+
+	mov	%esi, %edx
+
+	/* Copy trampoline code to the physical address 0. This is the place
```

Comment:
Why do we need to prepare trampoline code here? I suppose unit-test itself will have trampoline code prepared.

Reply:
The name may be misleading. The purpose of this piece is to move the whole binary to the desired place (i.e. 4M for acrn-unit-test binaries). The pref_address in zero_page is a recommendation to bootloaders but there is no guarantee that a bootloader will definitely put the image there. That's why I have this piece to copy the image to 4M.